src/layout.c

```diff
@@ -92,45 +92,56 @@ void position_heads(struct SList *heads) {
 }
 
 struct SList *order_heads(struct SList *order_name_desc, struct SList *heads) {
-	struct SList *heads_ordered = NULL;
-	struct Head *head;
-	struct SList *i, *j, *r;
+	if (!heads)
+		return NULL;
 
+	unsigned long n_order = slist_length(order_name_desc);
+	unsigned long i;
 	struct SList *sorting = slist_shallow_clone(heads);
 
-	// specified order - regex match
-	for (i = order_name_desc; i; i = i->nex) {
-		j = sorting;
-		while(j) {
-			head = j->val;
-			r = j;
-			j = j->nex;
-			if (!head) {
-				continue;
-			}
-			if (i->val && (
-						head_matches_name_desc_regex(i->val, head) ||
-						head_matches_name_desc_partial(i->val, head)
-						)) {
-				slist_append(&heads_ordered, head);
-				slist_remove(&sorting, &r);
-			}
-		}
+	// array of order to list of heads matched
+	struct SList **order_heads = calloc(n_order, sizeof(struct SList*));
+
+	// exact match
```

Comment:
Thanks for tagging me!

In my personal opinion, I much prefer the behaviour landed as part of my patch, where each line in the config is parsed exactly once (either using regex if preceded by "!" or a partial match otherwise).

For an order like: 

- eDP-1
- !^DP-[0-9]$
- eDP-2
- DP-2

with attached devices: DP-1, DP-2, DP-3, eDP-1, and eDP-2 (connected in this order); under this code the order will be:

- eDP-1
- eDP-2
- DP-2
- DP-1
- DP-3

In mine it would be:
- eDP-1
- DP-1
- DP-2
- DP-3
- eDP-2

where the final match of DP-2 has already been consumed by the regex. This latter order seems much more aligned with user intent.

If you wanted to be fancy about this, you could scan ahead in the list to check for explicit names, and if the currently tested name matches an exact name later in the list, then skip the regex match. However, I'm not sure I would be bothered to do this... I'd just document that the algorithm is greedy, and have people write more complex regex patterns as necessary.

Reply:
> * 
> 
> 
> In mine it would be:
> 
>     * eDP-1
> 
>     * DP-1
> 
>     * DP-2
> 
>     * DP-3
> 
>     * eDP-2

That looks correct; it's a bug.

Reply:
Doesn't my code already do that, then? I could be missing something?

Reply:
Oh... goodness. I just didn't read the code correctly. Sorry. I don't code in C as often as other languages.

For each item in order list, you create a new SList, and then add the matching items to the nested SList corresponding to that index; and then flatten it afterward. I think there's still something missing though because the regex can be applied multiple times... but let me go over it again >.<. Sorry!

Reply:
Looking over this again, it all looks good!

Reply:
The above actually came out as:

```
actual:
 .name = 'head0', .description = 'eDP-1',
 .name = 'head3', .description = 'DP-1',
 .name = 'head4', .description = 'DP-3',
 .name = 'head1', .description = 'eDP-2',
 .name = 'head2', .description = 'DP-2',
```

The exact match behaviour was explicitly requested #60 so I think I'll just have to leave it in, as it's the current behaviour.

> If you wanted to be fancy about this, you could scan ahead in the list to check for explicit names, and if the currently tested name matches an exact name later in the list, then skip the regex match.

That's what it's doing; it scans first for exact matches and pulls them out, then regexes, then partials.

Reply:
> Doesn't my code already do that, then? I could be missing something?

Yours was correct; I'm just fixing an existing bug with exact matches here.